### DISCUSSION TITLE
Allow durationfield to be represented as ISO-8601

### DISCUSSION THREAD

**Discussion by sevdog** [encode/django-rest-framework#8527] [Ideas & Suggestions]
As of now for date and time field there is a config to specify the output format which allows to use the ISO-8601 format (which is also supported by django in its parsers).
Whereas `DurationField` are **always** formatted using django base method `duration_string`.  There is no way to configure _a whole project_ to use `duration_iso_string` instead to have durations always formatted as ISO-8601.

Is it true that you can override the `DurationField` to use that format, but when you have a dependency which uses this kind of field this option would lead to overwrite that code, bringing maintenance problems.

It would be cool if there could be a settings option to switch between base and ISO representation of durations.

**Comment by lovelydinosaur** [comment]
Could you start by giving an example of what format you'd want to see?

date/times and durations are different types of things, and it's not clear to me that ISO-8601 is relevant here.

**Comment by sevdog** [comment]
@tomchristie I am referencing to the ISO-8601 which is also supported by [`parse_duration`](https://docs.djangoproject.com/en/3.2/ref/utils/#django.utils.dateparse.parse_duration) method:

> e.g. P4DT1H15M20S which is equivalent to 4 1:15:20

As of now the `DurationField` only allows the latter representation (`4 1:15:20`) in `to_representation` method: https://github.com/encode/django-rest-framework/blob/2506d0b4f2ac8bdbf35d33b3dd8a56f3e8d0da75/rest_framework/fields.py#L1404-L1405

Allowing to use ISO representation for duration  (via `django.utils.duration.duration_iso_string`)  would improve interoperability with other languages which may interact with provided API since this is standard format thus is easier to have a ready-to-use parser for it.

**Reply by lovelydinosaur** [reply]
Ah I see it now... https://en.wikipedia.org/wiki/ISO_8601#Durations

Since that's in line with the existing `DateField` and `TimeField` I think we'd accept a pull request adding that functionality.

A good place for someone to start on this would be proposing what the documentation would look like after the change, since that'd be a really good point of review. Then adding the functionality and test cases to match up with the updated docs.

**Comment by sevdog** [comment]
@tomchristie I have opened a PR #8532 (in draft) with the updated doc.

I have tried to keep it aligned with date-time fields, I'm waiting for your opinion.

**Comment by florianvazelle** [comment]
Hi, the feature now seems to be merged, great work ! :wave: 

I tested it and it revealed a limitation in Django (issue: https://code.djangoproject.com/ticket/36747). It seems that Django supports the ISO 8601 duration format, but cannot handle years and months (`P3Y`, `P2M`, `P1Y2M3DT4H` ...), which results in `None` being returned (and a `ValidationError` in `rest_framework`) instead of a `timedelta`.

Perhaps this should be noted in the documentation ?
